utilities/object_registry.cc

```diff
@@ -185,5 +207,29 @@ void ObjectRegistry::Dump(Logger *logger) const {
   }
 }
 
+Status ObjectRegistry::RegisterPlugin(const std::string &name,
+                                      const PluginFunc &plugin_func) {
+  Plugin plugin;
```

Comment:
If we have a public type called "Plugin" then the plugin itself should control creation of the "Plugin" object, right? That's not what we have here. Best I can tell, here the object you are calling "Plugin" is a holder for a glorified callback for the ObjectRegistry. And we have a callback for populating the holder of the glorified callback. Why so many layers?

Reply:
> 

At the moment, the Plugin contains the name of the plugin, its registrar function, and an argument field.  As I mentioned earlier, the struct previously contained more fields.

I could also envision the PluginFunc doing additional work.  For example, a plugin that supports OpenSSL encryption may load the OpenSSL library dynamically and make sure the functions are available.  If they are not, it might return an appropriate error message and not initialize registration functions.  Or it might still register functions that return an error if the factory method is called. 

Using just the ObjectRegistry call, there would be no way to the caller that Open SSL is not supported, except via the Factory method itself.

Reply:
> the Plugin contains the name of the plugin

Not currently

Reply:
> 

Right, it did and then I removed it as a duplication (there is also a name to the other methods).